### ISSUE TITLE
Exhaustive switch on a type union with a default case yields TS2339: Property 'x' does not exist on type 'never'

### ISSUE THREAD

**Issue by Gilead** [Microsoft/TypeScript#9838]
**TypeScript Version:**  2.0.0

**Code**

```ts
type TestType = 'foo' | 'bar';

interface TestInterface {
  type : TestType;
}

class Test {
  test = (param : TestInterface) : boolean => {
    switch (param.type) {
      case 'foo' : return true;
      case 'bar' : return false;
      default    : throw Error("Invalid type: "+ param.type); // TS2339: Property 'type' does not exist on type 'never'.
    }
  }
}

```

**Expected behavior:**
The code compiles with no errors.

**Actual behavior:** 
tsc test.ts
test.ts(14,53): error TS2339: Property 'type' does not exist on type 'never'.

I understand the compiler knows that with the *current* code the default case can never be reached and the type property can not have any possible valid TestType value in this case but the whole point of having a default case is to limit the scope of the human error, compiling with a different TS version, compiler flags etc. It can also be considered a safe coding practice.

The workaround is easy but ugly and should not be necessary:
```
default    : throw Error("Invalid type: "+ (<any>param).type);
```

**Comment by HerringtonDarkholme** [comment]
The code above cannot go wrong in TypeScript land, as you have noted. The only reasonable error I can conceive is someone pass a value from the outside of TS or someone naughtily assert `any`.

In any case, I think your workaround is a good alarm of `someone do wrong thing somewhere`, though.

**Comment by Gilead** [comment]
Indeed. Another example is a function called from a HTML template event handler with no type safety in which case you can just have a typo somewhere trigger the default case. This is why I believe TypeScript should allow seemingly unnecessary default cases.

**Comment by yahiko00** [comment]
I dislike this current behavior. This should not be an error. It is quite weird TypeScript could change the type of an object (to `never` or anything else).

**Comment by yortus** [comment]
If you can't trust that the passed in `param` is really a `TestInterface` at runtime, then you probably shouldn't assume it's an object with a `type` property. If for example `param` is actually `null` or `undefined`, then the `default` clause above will throw a `TypeError` trying to access the `type` property. If it's a string or number, then it will throw `"Invalid type: "`.

**Comment by RyanCavanaugh** [comment]
Seems wrong -- `param` should be of type `TestInterface` and `param.type` should be of type `never`

**Comment by RyanCavanaugh** [comment]
Seems to be fixed in Anders' current literal types PR

**Comment by ahejlsberg** [comment]
Fixed in #9407.
